modules/core/src/test/java/org/apache/ignite/session/SessionContextCacheInterceptorTest.java

```diff
@@ -0,0 +1,442 @@
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+import org.jetbrains.annotations.Nullable;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.junit.Assume.assumeFalse;
+import static org.junit.Assume.assumeTrue;
+
+/** */
+@RunWith(Parameterized.class)
+public class SessionContextCacheInterceptorTest extends GridCommonAbstractTest {
+    /** */
+    private static final int KEYS = 10;
+
+    /** */
+    @Parameterized.Parameter
+    public CacheAtomicityMode mode;
+
+    /** */
+    @Parameterized.Parameter(1)
+    public CacheWriteSynchronizationMode syncMode;
+
+    /** */
+    @Parameterized.Parameter(2)
+    public boolean clnNode;
+
+    /** */
+    @Parameterized.Parameter(3)
+    public boolean dynamicCache;
+
+    /** */
+    @Parameterized.Parameter(4)
+    public int backups;
+
+    /** */
+    private Ignite ign;
+
+    /** */
+    private IgniteCache<Integer, String> cache;
+
+    /** */
+    @Parameterized.Parameters(name = "mode={0}, syncMode={1}, clnNode={2}, dynamicCache={3}, backups={4}")
+    public static Collection<Object[]> params() {
+        List<Object[]> params = new ArrayList<>();
+
+        for (CacheAtomicityMode m: CacheAtomicityMode.values()) {
+            for (CacheWriteSynchronizationMode syncMode : CacheWriteSynchronizationMode.values()) {
+                for (boolean cln : new boolean[] {true, false}) {
+                    for (boolean dynCache : new boolean[] {true, false}) {
+                        if (m == CacheAtomicityMode.TRANSACTIONAL) {
+                            for (int backupCnt = 0; backupCnt <= 2; backupCnt++)
+                                params.add(new Object[] {m, syncMode, cln, dynCache, backupCnt});
+                        }
+                        else
+                            params.add(new Object[] {m, syncMode, cln, dynCache, 1});
+                    }
+                }
+            }
+        }
+
+        return params;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (!dynamicCache)
+            cfg.setCacheConfiguration(cacheConfig());
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        ign = startGrids(3);
+
+        if (clnNode)
+            ign = startClientGrid(3);
+
+        cache = ign.getOrCreateCache(cacheConfig());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
```

Comment:
Looks like we can create separate cache for each test mode.
This will allow to start grids only once - on test start.

You have to make test cache name constists of all params like 

`DEFAULT_CACHE_NAME + "_" + m + "_" + syncMode ...`